### PR TITLE
No replay

### PR DESCRIPTION
**Pull request by hoodmane**
Our package build process currently has a significant flaw: we first run `setup.py`, recording all compilation commands, then we rewrite these compilation commands to invoke `emcc` and replay them, and then we pray that the cross compiled executables ended up in the right place to go into the wheel. This is not a good strategy because the build script is allowed to implement arbitrary logic, and if it moves, renames, etc any of the output files then we lose track of them. This has repeatedly caused difficulty for us.

However, we also make no particularly significant use of the two pass approach. We can just do the simpler thing: capture the compiler commands as they occur, modify them as needed, and then run the fixed command.

I also added a patch to fix the numpy feature detection for wasm so that we don't have to include `_npyconfig.h` and `config.h`, numpy can generate them in the way it would for a native build. I opened a numpy PR that would fix the detection for us upstream:
https://github.com/numpy/numpy/pull/21154

This clears the way for us to switch to using pypa/build (as @henryiii has suggested) by removing our dependence on specific setuptools behavior.

This is on top of #2238.

### CODE REVIEW

File: packages/numpy/patches/0001-BUG-numpy-core-setup.py-math-feature-detection-on-wa.patch [hunk 1]
```diff
@@ -0,0 +1,271 @@
+From a315a30e116fe73f46a80eb1859dcf3d4a4ad4da Mon Sep 17 00:00:00 2001
```
Comment (hoodmane) [inline packages/numpy/patches/0001-BUG-numpy-core-setup.py-math-feature-detection-on-wa.patch:1]: This is https://github.com/numpy/numpy/pull/21154

**Review by ryanking13 — APPROVED** [review]
Thanks for your effort in working on this @hoodmane! This change will definitely make understanding our build system much easier.

The overall changes LGTM, please update docs [on our build system](https://pyodide.org/en/stable/development/new-packages.html), and the changelog update that briefly describes changes in our build system would be great.

File: pyodide-build/pyodide_build/buildpkg.py [hunk 2]
```diff
@@ -710,18 +660,18 @@ def build_package(
...
-    if not should_prepare_source and not srcpath.exists():
+    if continue_ and not srcpath.exists():
         raise OSError(
             "Cannot find source for rebuild. Expected to find the source "
             f"directory at the path {srcpath}, but that path does not exist."
         )
 
     with chdir(pkg_root), get_bash_runner() as bash_runner:
-        if should_prepare_source:
+        bash_runner.env["PKG_VERSION"] = version
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildpkg.py:670]: Right. I guess the long term plan is to try and make our `meta.yaml` system work close to the same way as conda, so we might try to make sure that these variables etc match the way conda works. I think in the short term I can put a unstable warning on the build docs saying that the build system is actively changing (maybe @rth already added a warning to that effect).

File: pyodide-build/pyodide_build/buildpkg.py [hunk 3]
```diff
@@ -656,7 +608,9 @@ def needs_rebuild(
 
     def source_files():
         yield pkg_root / "meta.yaml"
-        yield from source_metadata.get("patches", [])
+        yield from (
+            pkg_root / patch_path for patch_path in source_metadata.get("patches", [])
+        )
         yield from (x[0] for x in source_metadata.get("extras", []))
         src_path = source_metadata.get("path")
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildpkg.py:615]: I guess so. I wonder whether this was already broken or if I broke it in this PR somehow.

File: pyodide-build/pyodide_build/buildpkg.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildpkg.py:675]: I documented a few of these, but I think for instance the situation with `HOSTINSTALLDIR` will change already in #2272 so I am reluctant to document it just yet.

**Comment by hoodmane** [comment]
Thanks for the review @ryanking13!

**Review by rth — COMMENTED** [review]
Thanks for working on this @hoodmane, it looks very neat! Also thanks for the reviews @ryanking13 !

If there are no two stage builds anymore, wouldn't that mean that https://pyodide.org/en/latest/development/new-packages.html#partial-rebuilds need updating?

Also I don't really understand how building packages for host happens with this PR (e.g. numpy installed on host). We do have to run the build once for each arch don't we?
Maybe we could use this opportunity to document how cross-compilation works in broad terms in Pyodide, possibly by adding a page on "Build system" with a "Cross-compilation" subsection nested under https://pyodide.org/en/latest/development/new-packages.html (same as "The meta.yaml specification")  or otherwise in the top level of "Development"?

**Pull request merged**